data/evidence/store.go

```diff
@@ -0,0 +1,250 @@
+}
+
+func (es *EvidenceStore) IterateRequests(fn func(ar *AllegationRequest) bool) (stopped bool) {
+	key := []byte("_ark_")
+	prefixKey := append(es.prefix, key...)
+	return es.state.IterateRange(
+		prefixKey,
+		storage.Rangefix(string(prefixKey)),
+		true,
+		func(key, value []byte) bool {
+			ar, err := (&AllegationRequest{}).FromBytes(value)
+			if err != nil {
+				fmt.Println("failed to deserialize allegation request")
+				return false
+			}
+			return fn(ar)
+		},
+	)
+}
+
+func (es *EvidenceStore) CreateSuspiciousValidator(validatorAddress keys.Address, status int8, height int64, createdAt *time.Time) (*LastValidatorHistory, error) {
+	lvh := NewLastValidatorHistory(validatorAddress, status, height, createdAt)
+	err := es.UpdateSuspiciousValidator(lvh)
+	return lvh, err
+}
+
+func (es *EvidenceStore) UpdateSuspiciousValidator(lvh *LastValidatorHistory) error {
+	dat, err := serialize.GetSerializer(serialize.PERSISTENT).Serialize(lvh)
+	if err != nil {
+		return err
+	}
+	return es.Set(es.getSuspiciousValidatorKey(lvh.Address), dat)
+}
+
+func (es *EvidenceStore) IsFrozenValidator(validatorAddress keys.Address) bool {
+	dat, err := es.Get(es.getSuspiciousValidatorKey(validatorAddress))
+	if err != nil {
+		return false
+	}
+	if len(dat) == 0 {
+		return false
+	}
+	lvh := &LastValidatorHistory{}
+	err = serialize.GetSerializer(serialize.PERSISTENT).Deserialize(dat, lvh)
+	if err != nil {
+		return false
+	}
+
+	if err != nil {
+		return false
+	}
+	return lvh.IsFrozen()
+}
+
+func (es *EvidenceStore) HandleRelease(options *Options, validatorAddress keys.Address, blockHeight int64, blockCreatedAt time.Time) error {
+	lvh, err := es.GetSuspiciousValidator(validatorAddress, 0, 0)
+	if err != nil {
+		return err
+	}
+	if !lvh.IsFrozen() {
+		return fmt.Errorf("Validator \"%s\" already released", validatorAddress)
+	}
+
+	isReady, err := lvh.ReleaseReady(options, blockCreatedAt)
+	if err != nil {
+		return err
+	}
+
+	if !isReady {
+		return fmt.Errorf("Validator \"%s\" not ready for release", validatorAddress)
+	}
+
+	lvh.ReleaseHeight = blockHeight
+	lvh.ReleaseAt = &blockCreatedAt
+	err = es.UpdateSuspiciousValidator(lvh)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (es *EvidenceStore) PerformAllegation(validatorAddress keys.Address, maliciousAddress keys.Address, ID string, blockHeight int64, proofMsg string) error {
+	es.mux.Lock()
+	defer es.mux.Unlock()
+
+	isBusy := es.IsRequestIDBusy(ID)
+	if isBusy {
+		return fmt.Errorf("request ID %s already handled\n", ID)
+	}
+
+	ar := NewAllegationRequest(ID, validatorAddress, maliciousAddress, blockHeight, proofMsg)
+	err := es.SetAllegationRequest(ar)
+	if err != nil {
+		return err
+	}
+
+	at, err := es.GetAllegationTracker()
+	if err != nil {
+		return err
+	}
+
+	at.Requests[ID] = true
+
+	err = es.SetAllegationTracker(at)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (es *EvidenceStore) Vote(requestID string, voteAddress keys.Address, choice int8) error {
```

Comment:
Have you implemented dumping store data to genesis.json?

Reply:
@ws4charlie what do you mean by dumping store?

Reply:
> @ws4charlie what do you mean by dumping store?

Yes. Skip this if it's not in requirement doc.